File: ML/practicing codes/Reinforcement Learning/Thompson Algorithm/Thompson_sampling.py
# Thompson Sampling Algorithm

# Import the libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Import the dataset
dataset = pd.read_csv("Ads_CTR_Optimisation.csv")

# Build the Algorithm
import random
N = 10000
d = 10
numbers_of_zero_rewards = [0] * d
numbers_of_one_rewards = [0] * d
selected_ads = []
total_reward = 0

for n in range(0, N):
    max_random = 0
    ad = 0
    for i in range(0, d):
        random_beta = random.betavariate(numbers_of_one_rewards[i] + 1, numbers_of_zero_rewards[i] + 1)
        if random_beta > max_random:
            max_random = random_beta
            ad = i
    selected_ads.append(ad)
    reward = dataset.values[n, ad]
    if reward == 1:
        numbers_of_one_rewards[ad] = numbers_of_one_rewards[ad] + 1
    else:
        numbers_of_zero_rewards[ad] = numbers_of_zero_rewards[ad] + 1
    total_reward += reward
        
# Visualise the results
plt.hist(selected_ads)
plt.title("The Histogram of Selected Ads")
plt.xlabel("Ads")
plt.ylabel("Number Of Selections")

File: ML/practicing codes/Reinforcement Learning/Thompson Algorithm/Thompson_sampling.R
# Thompson Sampling Algorithm

# Import the dataset
dataset = read.csv("Ads_CTR_Optimisation.csv")

# Build the model
N = 10000
d = 10
number_of_one_rewards = integer(d)
number_of_zero_rewards = integer(d)
selected_ads = integer()
total_rewards = 0

for (n in 1:N) {
  max_random = 0
  ad = 0
  for (i in 1:d) {
    random_beta = rbeta(n = 1,
                        shape1 = number_of_one_rewards[i] + 1,
                        shape2 = number_of_zero_rewards[i] + 1)
    if (random_beta > max_random) {
      max_random = random_beta
      ad = i
    }
  }
  
  selected_ads = append(selected_ads, ad)
  reward = dataset[n, ad]
  if (reward == 1) {
    number_of_one_rewards[ad] = number_of_one_rewards[ad] + 1
  } else {
    number_of_zero_rewards[ad] = number_of_zero_rewards[ad] + 1
  }
    
  total_rewards = total_rewards + reward
}

# visualise the results
hist(selected_ads,
     col = 'purple',
     main = "Histogram of the Selected Ads",
     xlab = "Ads",
     ylab = "Number of each Ad selections")


